core/templates/dev/head/domain/topic/TopicObjectFactory.js

```diff
@@ -0,0 +1,266 @@
+      this._name = name;
+    };
+
+    Topic.prototype.getDescription = function() {
+      return this._description;
+    };
+
+    Topic.prototype.getNextSubtopicId = function() {
+      return this._nextSubtopicId;
+    };
+
+    Topic.prototype.setDescription = function(description) {
+      this._description = description;
+    };
+
+    Topic.prototype.getLanguageCode = function() {
+      return this._languageCode;
+    };
+
+    Topic.prototype.setLanguageCode = function(languageCode) {
+      this._languageCode = languageCode;
+    };
+
+    Topic.prototype.getVersion = function() {
+      return this._version;
+    };
+
+    Topic.prototype.getSubtopicById = function(subtopicId) {
+      for (var i = 0; i < this._subtopics.length; i++) {
+        var id = this._subtopics[i].getId();
+        if (id === subtopicId) {
+          return this._subtopics[i];
+        }
+      }
+      return false;
+    };
+
+    // The following function is only meant to be called from the undo of
+    // a delete subtopic operation and not anywhere else. For adding subtopics
+    // in general, use topic.addSubtopic()
+    Topic.prototype.undoDeleteSubtopic = function(id, title, skillIds) {
+      var newSubtopic = SubtopicObjectFactory.createFromTitle(id, title);
+      for (var i = 0; i < skillIds.length; i++) {
+        newSubtopic.addSkillId(skillIds[i]);
+      }
+      this._subtopics.push(newSubtopic);
+      this._nextSubtopicId++;
+    };
+
+    // Adds a new frontend subtopic domain object to this topic.
+    Topic.prototype.addSubtopic = function(title) {
+      var newSubtopic = SubtopicObjectFactory.createFromTitle(
+        this._nextSubtopicId, title);
+      this._subtopics.push(newSubtopic);
+      this._nextSubtopicId++;
+    };
+
+    // Attempts to remove a subtopic from this topic given the
+    // subtopic ID.
+    Topic.prototype.deleteSubtopic = function(subtopicId) {
+      for (var i = 0; i < this._subtopics.length; i++) {
+        if (this._subtopics[i].getId() === subtopicId) {
+          // When a subtopic is deleted, all the skills in it are moved to
+          // uncategorized skill ids.
+          var skillIds = this._subtopics[i].getSkillIds();
+          for (var j = 0; j < skillIds.length; j++) {
+            var skillId = skillIds[j];
+            if (this._uncategorizedSkillIds.indexOf(skillId) === -1) {
+              this._uncategorizedSkillIds.push(skillId);
+            }
+          }
+          this._subtopics.splice(i, 1);
+          return true;
+        }
+      }
+      return false;
+    };
+
+    Topic.prototype.clearSubtopics = function() {
+      this._subtopics.length = 0;
+    };
+
+    Topic.prototype.getSubtopics = function() {
+      return this._subtopics.slice();
+    };
+
+    Topic.prototype.addCanonicalStoryId = function(storyId) {
+      if (this._canonicalStoryIds.indexOf(storyId) === -1) {
+        this._canonicalStoryIds.push(storyId);
+        return true;
```

Comment:
This was done to prevent cases when adding an already existing story, or removing an invalid skill, but I am also checking it in TopicUpdateServices, will remove these.

Reply:
Done.